django/db/models/fields/related_descriptors.py

```diff
@@ -1080,11 +1058,11 @@ def _add_items(self, source_field_name, target_field_name, *objs):
 
                     # Add the ones that aren't there already
                     self.through._default_manager.using(db).bulk_create([
-                        self.through(**{
+                        self.through(**dict(extra_values, **{
                             '%s_id' % source_field_name: self.related_val[0],
                             '%s_id' % target_field_name: obj_id,
-                        })
-                        for obj_id in new_ids
+                        }))
+                        for obj_id, extra_values in zip(new_ids, intermediate_values)
```

Comment:
keep in mind new_ids could be smaller than objs, messing up index order so intermediate_values won't match up, right? Also, new_ids is a _set_, so order is lost.

Reply:
Yep, missed both those points. Clearly needs a rework....